xrootd/client/client.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+/*
+Package client implements the XRootD client following protocol from http://xrootd.org.
+
+The NewClient function connects to a server:
+
+	ctx := context.Background()
+
+	client, err := NewClient(ctx, *Addr)
+	if err != nil {
+		// handle error
+	}
+
+	// ...
+
+	if err := client.Close(); err != nil {
+		// handle error
+	}
+*/
+package client // import "go-hep.org/x/hep/xrootd/client"
+
+import (
+	"context"
+	"io"
+	"net"
+
+	"go-hep.org/x/hep/xrootd/encoder"
+	"go-hep.org/x/hep/xrootd/mux"
+	"go-hep.org/x/hep/xrootd/protocol"
+)
+
+// A Client to xrootd server which allows to send requests and receive responses.
+// Concurrent requests are supported.
+// Zero value is invalid, Client should be instantiated using NewClient.
+type Client struct {
+	cancel          context.CancelFunc
+	conn            net.Conn
+	mux             *mux.Mux
+	protocolVersion int32
+}
+
+// NewClient creates a new xrootd client that connects to the given address.
+// When the context expires, a response handling is stopped, however, it is
+// necessary to call Cancel to correctly free resources.
+func NewClient(ctx context.Context, address string) (*Client, error) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	var d net.Dialer
+	conn, err := d.DialContext(ctx, "tcp", address)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &Client{cancel, conn, mux.New(), 0}
+
+	go client.consume(ctx)
+
+	if err := client.handshake(ctx); err != nil {
+		client.Close()
+		return nil, err
+	}
+
+	return client, nil
+}
+
+// Close closes the connection. Any blocked operation will be unblocked and return error.
+func (client *Client) Close() error {
+	client.cancel()
+
+	err := client.conn.Close()
+	if err != nil {
+		return err
+	}
+
+	client.mux.Close()
```

Comment:
shouldn't we always close the multiplexer, even if `client.conn.Close` failed ?
ie:

```go
err := client.conn.Close()
client.mux.Close()
return err
```

or, even:
```go
client.mux.Close()
return client.conn.Close()
```

Reply:
My bad. Thank you!